### DISCUSSION TITLE
Allow overriding certificate hash algorithm

### DISCUSSION THREAD

**Discussion by bagasme** [smallstep/certificates#1266] [Ideas]
Currently hash algorithm used to create and sign certificates is determined by hash algorithm in the CA certificate. For example, a certificate signed by default P-256 curve CA certificate uses SHA-256 and SHA-384 for certificates signed by P-384 CA.

`step ca` and `step certificate` should have an option to select desired hash algorithm (`--hash`). The valid values for the option shall be
`SHA-256`, `SHA-384`, `SHA-512`, and `default` (CA-dependent as before).

**Comment by tashian** [comment]
Hi @bagasme, [I've created an issue from this discussion](https://github.com/smallstep/certificates/issues/1272).

**Comment by maraino** [comment]
@bagasme @tashian For `step certificate` you should be able to create a certificate adding your preferred `signatureAlgorithm` for example:
```tpl
{
  "subject": {{ toJson .Subject }},
  "issuer": {{ toJson .Subject }},
  "keyUsage": ["certSign", "crlSign"],
  "basicConstraints": {
    "isCA": true,
    "maxPathLen": 1
  },
  "signatureAlgorithm": "ECDSA-SHA512"
}
```

And in `step ca` you can always define a provisioner template like that:
```tpl
{
  "subject": {{ toJson .Subject }},
  "sans": {{ toJson .SANs }},
{{- if typeIs "*rsa.PublicKey" .Insecure.CR.PublicKey }}
  "keyUsage": ["keyEncipherment", "digitalSignature"],
{{- else }}
  "keyUsage": ["digitalSignature"],
{{- end }}
  "extKeyUsage": ["serverAuth", "clientAuth"],
  "signatureAlgorithm": "ECDSA-SHA384"
}
```

I would not recommend it, but you can even define the signature algorithm from the command line if you change it to `"signatureAlgorithm": {{ toJson .Insecure.User.signatureAlgorithm }}`, and then just:
```
step ca certificate --set "signatureAlgorithm=ECDSA-SHA512" localhost localhost.crt localhost.key
```

Of course, it will fail if your intermediate key doesn't support it, for example, you won't be able to sign using `SHA256-RSAPSS` if your intermediate key is an EC key.
